Audit3. Convert random buckets selection to user defined buckets.

shamil-gadelshin opened the pull request:
We remove a random selection of the storage and distribution buckets on dynamic bags creation from the storage pallet and replace it with user-defined buckets.

#### Changes
- removed a random selection of the buckets
- added new parameters (buckets) to the bag creation methods
- modified extrinsics for dynamic bag creation policy modifications (enforced new checks for parameters)
- fixed tests for storage and content pallets.

[Original issue](https://github.com/Joystream/joystream/issues/3289)

shamil-gadelshin committed:
runtime: Change buckets assignment on bag creation.

Changes:
- storage and content pallets
- removed random storage and distribution bucket assignment
- introduced storage and distribution buckets parameters on the dynamic bag creation
- introduced storage buckets checks

shamil-gadelshin committed:
runtime: storage: Add distribution buckets checks.

- add distribution buckets checks on dynamic bag creation.
- remove obsolete trait fields MaxRandomIterationNumber and Randomness

shamil-gadelshin committed:
runtime: content: Fix tests.

shamil-gadelshin committed:
runtime: storage: Add constraints for bag creation policy.

shamil-gadelshin committed:
runtime: content: Fix tests.

shamil-gadelshin commented:
We need to update the `audit3` branch with the latest `olympia` code in order to review this PR properly.

dobertRowneySr commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -2978,6 +2958,25 @@ impl<T: Trait> Module<T> {
             Self::check_buckets_for_overflow(storage_buckets, &bag_change.voucher_update)?;
         }
 
+        // Validate distribution bucket IDs
```
This has become a long function now with several sub validation routines. Can you please refactor each unit with self contained validation logic (like distribution bucket validation) into its own separate function. I think it is clearer that way

shamil-gadelshin replied:
Fixed here: https://github.com/Joystream/joystream/pull/3348/commits/32d252427ebd6c3b417662badd523e572ab3cbc4

dobertRowneySr reviewed (changes requested):
It's ok. No major issue found. I just left a small request in order to enhance code clarity

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-user-defined-buckets

# Conflicts:
#	runtime-modules/content/src/lib.rs
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/fixtures.rs
#	runtime-modules/content/src/tests/videos.rs
#	runtime-modules/content/src/types.rs
#	runtime-modules/storage/src/lib.rs
#	runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs
#	runtime-modules/storage/src/tests/fixtures.rs
#	runtime-modules/storage/src/tests/mod.rs

shamil-gadelshin committed:
runtime: storage: Fix review comments.

- refactor storage and distribution buckets validation on  the dynamic bag creation.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-user-defined-buckets

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/tests/fixtures.rs
#	runtime-modules/content/src/tests/videos.rs

dobertRowneySr reviewed (approved):
LGTM
